### PR TITLE
Fix various edge cases in the unlocalized body calculation

### PR DESCRIPTION
**Pull request by nico-famedly**
See the additional test cases. There are still some weird edge cases, but those are documented in the tests now.

### CODE REVIEW

File: test/event_test.dart [hunk 1]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+        isEdit: true,
+        editBody: null,
+        editFormattedBody: null,
+        editHtml: true,
+      );
+      testUnlocalizedBody(
+        expectation: '**formatted body**',
+        plaintextBody: true,
+        body: 'body',
+        formattedBody: '<b>formatted body</b>',
```
Comment (td-famedly) [inline test/event_test.dart:1433]: same here, would expect editBody

File: test/event_test.dart [hunk 2]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+        editFormattedBody: null,
+        editHtml: false,
+      );
+      testUnlocalizedBody(
+        expectation: '**formatted body**',
+        plaintextBody: true,
+        body: 'body',
+        formattedBody: '<b>formatted body</b>',
+        html: true,
+        isEdit: true,
```
Comment (td-famedly) [inline test/event_test.dart:1446]: same here

File: test/event_test.dart [hunk 3]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+      );
+      testUnlocalizedBody(
+        expectation: 'edit body',
+        plaintextBody: false,
+        body: 'body',
+        formattedBody: '<b>formatted body</b>',
+        html: true,
+        isEdit: true,
+        editBody: 'edit body',
+        editFormattedBody: '<b>edit formatted body</b>',
```
Comment (td-famedly) [inline test/event_test.dart:1507]: hm shouldn't this return the edited formatted body?

File: test/event_test.dart [hunk 4]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+        body: 'body',
+        formattedBody: '<b>formatted body</b>',
+        html: false,
+        isEdit: true,
+        editBody: 'edit body',
+        editFormattedBody: null,
+        editHtml: true,
+      );
+      testUnlocalizedBody(
+        expectation: 'edit body',
```
Comment (td-famedly) [inline test/event_test.dart:1533]: again would expect '5'

File: lib/src/event.dart [hunk 5]
```diff
@@ -292,15 +292,18 @@ class Event extends MatrixEvent {
...
   }
 
   /// Use this to get a plain-text representation of the event, stripping things
   /// like spoilers and thelike. Useful for plain text notifications.
-  String get plaintextBody => content['format'] == 'org.matrix.custom.html'
-      ? HtmlToText.convert(formattedText)
-      : body;
+  String get plaintextBody => switch (formattedText) {
+        // if the formattedText is empty, fallback to body
+        '' => body,
```
[the remark this answers is not in the document]
Comment (nico-famedly) [reply lib/src/event.dart:302]: We already discussed that, and no, I think that is harder to read.

File: test/event_test.dart [hunk 6]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+        formattedBody: '<b>formatted body</b>',
+        html: false,
+        isEdit: false,
+        editBody: null,
+        editFormattedBody: null,
+        editHtml: false,
+      );
+      testUnlocalizedBody(
+        // not sure we actually want m.room.message here and not an empty string
+        expectation: 'm.room.message',
```
[the remark this answers is not in the document]
Comment (nico-famedly) [reply test/event_test.dart:1308]: Well, that is historical behaviour. At least some events don't have a body, how do you localize them? All of them just as empty?

File: test/event_test.dart [hunk 7]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+        html: false,
+        isEdit: false,
+        editBody: null,
+        editFormattedBody: null,
+        editHtml: false,
+      );
+      testUnlocalizedBody(
+        expectation: 'm.room.message',
+        plaintextBody: false,
+        body: 5,
```
[the remark this answers is not in the document]
Comment (nico-famedly) [reply test/event_test.dart:1332]: Why? A number is not a string, so it falls back to using the same logic as if the body was missing?

File: test/event_test.dart (hunk 3 above)
Comment (nico-famedly) [reply test/event_test.dart:1507]: No, because it has plaintextBody set to false.

File: test/event_test.dart (hunk 1 above)
Comment (nico-famedly) [reply test/event_test.dart:1433]: Edit body is null, which is an invalid value, so it falls back to using the event formatted body. So you expect an empty string here? Even though the newContent is missing?

File: test/event_test.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (td-famedly) [reply test/event_test.dart:1308]: localization could just use a the event type if the body is empty, inserting the type when the body is empty is very confusing imo

File: test/event_test.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (td-famedly) [reply test/event_test.dart:1332]: Yes but we could be more flexible here? I don't see why not just show the int content if we can

File: test/event_test.dart (hunk 1 above)
Comment (td-famedly) [reply test/event_test.dart:1433]: > So you expect an empty string here?

yep, falling back to the old event sounds bad

Comment (nico-famedly) [reply test/event_test.dart:1433]: It's not the old event, it is the fallback body in the new event, it just usually has a weird `*` in front of it or so.

File: test/event_test.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (nico-famedly) [reply test/event_test.dart:1308]: We can change that, but that would be something @krille-chan signs off on, imo.

File: test/event_test.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (td-famedly) [reply test/event_test.dart:1332]: hm sg

File: test/event_test.dart [hunk 8]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+        isEdit: false,
+        editBody: null,
+        editFormattedBody: null,
+        editHtml: false,
+      );
+      testUnlocalizedBody(
+        expectation: 'body',
+        plaintextBody: false,
+        body: 'body',
+        formattedBody: '<b>formatted body</b>',
```
[the remark this answers is not in the document]
Comment (td-famedly) [reply test/event_test.dart:1411]: sg

File: test/event_test.dart (hunk 1 above)
Comment (td-famedly) [reply test/event_test.dart:1433]: ah right that makes sense

**Review by td-famedly — COMMENTED** [review]
lg

File: test/event_test.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (td-famedly) [reply test/event_test.dart:1308]: moved to https://github.com/famedly/product-management/issues/2506

File: test/event_test.dart [hunk 9]
```diff
@@ -1200,6 +1212,581 @@ void main() {
...
+        body: '',
+        formattedBody: '<b>formatted body</b>',
+        html: false,
+        isEdit: false,
+        editBody: null,
+        editFormattedBody: null,
+        editHtml: false,
+      );
+      testUnlocalizedBody(
+        expectation: 'm.room.message',
```
[the remark this answers is not in the document]
Comment (td-famedly) [reply test/event_test.dart:1319]: moved to https://github.com/famedly/product-management/issues/2506

File: lib/src/event.dart [hunk 10]
```diff
@@ -851,37 +857,44 @@ class Event extends MatrixEvent {
...
     if (redacted) {
       return 'Removed by ${senderFromMemoryOrFallback.displayName ?? senderId}';
     }
     var body = plaintextBody ? this.plaintextBody : this.body;
 
-    // we need to know if the message is an html message to be able to determine
-    // if we need to strip the reply fallback.
-    var htmlMessage = content['format'] != 'org.matrix.custom.html';
+    // Html messages will already have their reply fallback removed during the Html to Text conversion.
+    var mayHaveReplyFallback = !plaintextBody ||
```
[the remark this answers is not in the document]
Comment (krille-chan) [reply lib/src/event.dart:872]: I thought we want to avoid declaring types when they can be calculated from the given value?

[the remark this answers is not in the document]
Comment (td-famedly) [reply lib/src/event.dart:872]: I don't think we enabled those lints yet 🤔

[the remark this answers is not in the document]
Comment (td-famedly) [reply lib/src/event.dart:872]: yep still in refinement https://github.com/famedly/product-management/issues/2189 I think we were waiting for you to come back for disucssion

[the remark this answers is not in the document]
Comment (nico-famedly) [reply lib/src/event.dart:872]: The return value is already a bool, so why would I need to spell that out? What would be fixed by being explicit here?

[the remark this answers is not in the document]
Comment (td-famedly) [reply lib/src/event.dart:872]: just because it looks nicer and more consistent with our other approach iirc. But ofc a nitpick feel free to ignore and resolve this thread

[the remark this answers is not in the document]
Comment (nico-famedly) [reply lib/src/event.dart:872]: I don't think it is consistent with anything in that file and afaik the linter yells at me in many cases when I do spell out the type :D

So I'm just going to ignore this.

**Review by td-famedly — APPROVED** [review]
probably fine
